### ISSUE TITLE
NaN loss in indoor training

### ISSUE THREAD

**Issue by MeNinaa** [chenhsuanlin/spatial-transformer-GAN#22]
Hi,

thanks for sharing your code. I tested it with the indoor pre trained model, and it works fine, however when I try to train it eventually the loss starts returning NaN. When I actually get numerical values, they are very high. Any ideas why this happens?

Thanks

spatial-transformer-GAN-master/indoor$ ./train.sh
=======================================================
pretrain_homo.py (pretrain STN with homography perturbation)
=======================================================
(0) pretrain
------------------------------------------
GPU device: 0, batch size: 20, warps: 1
image size: 120x160
[GP] stddev=1e-02, lr=1e-04, decay=1.0, step=10000, update=1
[D]  stddev=1e-02, lr=1e-04, decay=1.0, step=10000, update=2
------------------------------------------
training model (0) pretrain...
building graph...
loading training data...
loading test data...
======= TRAINING START =======
2019-06-10 16:33:44.057315: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-06-10 16:33:44.187251: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:897] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2019-06-10 16:33:44.187628: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1392] Found device 0 with properties: 
name: GeForce GTX 1080 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.62
pciBusID: 0000:01:00.0
totalMemory: 10.91GiB freeMemory: 9.55GiB
2019-06-10 16:33:44.187641: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1471] Adding visible gpu devices: 0
2019-06-10 16:33:44.362210: I tensorflow/core/common_runtime/gpu/gpu_device.cc:952] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-06-10 16:33:44.362279: I tensorflow/core/common_runtime/gpu/gpu_device.cc:958]      0 
2019-06-10 16:33:44.362286: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] 0:   N 
2019-06-10 16:33:44.370289: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1084] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 9236 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:01:00.0, compute capability: 6.1)
start training...
it.20/200000 lr=1e-04 loss=2519.5112 time=3.22
it.40/200000 lr=1e-04 loss=72730792.0000 time=3.84
it.60/200000 lr=1e-04 loss=142534492160.0000 time=4.38
it.80/200000 lr=1e-04 loss=56200574009344.0000 time=4.92
it.100/200000 lr=1e-04 loss=7838565497044992.0000 time=5.46
it.120/200000 lr=1e-04 loss=391058524647456768.0000 time=5.99
it.140/200000 lr=1e-04 loss=13523502639458811904.0000 time=6.55
it.160/200000 lr=1e-04 loss=258345773254062374912.0000 time=7.09
it.180/200000 lr=1e-04 loss=nan time=7.63
it.200/200000 lr=1e-04 loss=nan time=8.17
it.220/200000 lr=1e-04 loss=nan time=8.71
it.240/200000 lr=1e-04 loss=nan time=9.26
it.260/200000 lr=1e-04 loss=nan time=9.80
it.280/200000 lr=1e-04 loss=nan time=10.35
it.300/200000 lr=1e-04 loss=nan time=10.92
it.320/200000 lr=1e-04 loss=nan time=11.48
it.340/200000 lr=1e-04 loss=nan time=12.01
it.360/200000 lr=1e-04 loss=nan time=12.55
it.380/200000 lr=1e-04 loss=nan time=13.08
it.400/200000 lr=1e-04 loss=nan time=13.66
it.420/200000 lr=1e-04 loss=nan time=14.22
it.440/200000 lr=1e-04 loss=nan time=14.76
it.460/200000 lr=1e-04 loss=nan time=15.33
it.480/200000 lr=1e-04 loss=nan time=15.87
it.500/200000 lr=1e-04 loss=nan time=16.41
it.520/200000 lr=1e-04 loss=nan time=16.95
it.540/200000 lr=1e-04 loss=nan time=17.49
it.560/200000 lr=1e-04 loss=nan time=18.02
it.580/200000 lr=1e-04 loss=nan time=18.59
it.600/200000 lr=1e-04 loss=nan time=19.14
it.620/200000 lr=1e-04 loss=nan time=19.67
it.640/200000 lr=1e-04 loss=nan time=20.20
it.660/200000 lr=1e-04 loss=nan time=20.75
it.680/200000 lr=1e-04 loss=nan time=21.29
it.700/200000 lr=1e-04 loss=nan time=21.83
it.720/200000 lr=1e-04 loss=nan time=22.37
it.740/200000 lr=1e-04 loss=nan time=22.91
it.760/200000 lr=1e-04 loss=nan time=23.45
it.780/200000 lr=1e-04 loss=nan time=23.99
it.800/200000 lr=1e-04 loss=nan time=24.54
it.820/200000 lr=1e-04 loss=nan time=25.09
it.840/200000 lr=1e-04 loss=nan time=25.64
it.860/200000 lr=1e-04 loss=nan time=26.19
it.880/200000 lr=1e-04 loss=nan time=26.74
it.900/200000 lr=1e-04 loss=nan time=27.27
it.920/200000 lr=1e-04 loss=nan time=27.82
it.940/200000 lr=1e-04 loss=nan time=28.37
it.960/200000 lr=1e-04 loss=nan time=28.90
it.980/200000 lr=1e-04 loss=nan time=29.43

**Comment by chenhsuanlin** [comment]
I don't think I've seen exploding losses in the pretraining phase, nor could I reproduce it. Could you try another run (in case you just very unluckily hit a bad random seed), or perhaps try from a clean clone of the repo?

**Comment by MeNinaa** [comment]
Is it supposed to be like this?
it.20/200000 lr=1e-04 loss=0.7875 time=2.67
it.40/200000 lr=1e-04 loss=0.6425 time=3.31
it.60/200000 lr=1e-04 loss=0.5574 time=3.95
it.80/200000 lr=1e-04 loss=0.7066 time=4.56
it.100/200000 lr=1e-04 loss=0.7726 time=5.18
it.120/200000 lr=1e-04 loss=0.7129 time=5.80
it.140/200000 lr=1e-04 loss=0.6628 time=6.41
it.160/200000 lr=1e-04 loss=0.7090 time=7.03
it.180/200000 lr=1e-04 loss=0.5476 time=7.64
it.200/200000 lr=1e-04 loss=0.7959 time=8.23
it.220/200000 lr=1e-04 loss=0.7221 time=8.83
it.240/200000 lr=1e-04 loss=0.5513 time=9.45
it.260/200000 lr=1e-04 loss=0.6582 time=9.99
it.280/200000 lr=1e-04 loss=0.6386 time=10.53
it.300/200000 lr=1e-04 loss=0.6173 time=11.11
it.320/200000 lr=1e-04 loss=0.7147 time=11.68
it.340/200000 lr=1e-04 loss=0.5388 time=12.22
it.360/200000 lr=1e-04 loss=0.7596 time=12.78
it.380/200000 lr=1e-04 loss=0.6207 time=13.34
it.400/200000 lr=1e-04 loss=0.5672 time=13.88
it.420/200000 lr=1e-04 loss=0.5410 time=14.42


Also, how many epochs should I run to obtain similar test results please?

**Comment by MeNinaa** [comment]
Sorry, just saw that there is an upper limit.. But are the losses supposed to be like that?

Thanks

**Comment by chenhsuanlin** [comment]
Yes, this looks about right. If you look at Tensorboard, you should also see a decreasing trend of the loss.
Please feel free to reopen if you have further issues.
